### PR TITLE
SDK Architecture document

### LINKED ISSUE #146: Documentation: References > Architecture
Should include
- Flow charts/diagrams
- How docker images/base images are generated and used
- Collector Framework 2 API specification

### LINKED ISSUE #148: Document commands.cfg file

### PR DESCRIPTION
**Pull request by kjrokos**
Draft

Information about the architecture of a containerized adapter.

Please pay special attention to:
* Organization of data
* Clarity of data
* Omissions

Resolves #146 
Resolves #148

### CODE REVIEW

File: doc/architecture.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+Architecture
+============
+
+This section describes how a Management Pack works, and how it interacts with the VMware 
+Aria Operations Collector.
+
+## Management Pack
+A management pack consists of content, metadata, and an adapter.
+
+* Content includes Dashboards, Reports, Icons, Symptoms, Alerts, Traversals, others.
```
Comment (MarkArbogast) [inline doc/architecture.md:10]: Each type of content could be a hyperlink to an explanation of what it is. 

Same with describe.xml and manifest.xml.

File: doc/architecture.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+* Metadata includes the describe.xml file that describes an adapter, and the 
+  `manifest.txt` file that describes the management pack as a whole.
+* The Adapter is the code that performs the collection and sends the data to VMware Aria
```
Comment (MarkArbogast) [inline doc/architecture.md:13]: Might be worth pointing out that the Adapter code is the main part that a developer needs to write.

File: doc/architecture.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+* The Adapter is the code that performs the collection and sends the data to VMware Aria 
+  Operations.
+
+### Adapter Container Image
+
+The adapter container image consists of two parts:
+* A REST server specified by [Collector Framework 2](../vmware_aria_operations_integration_sdk/api/vmware-aria-operations-collector-fwk2.json).
+* The adapter code that performs the collection.
+
+In an adapter created using the SDK, the REST server is provided inside a base container
```
Comment (MarkArbogast) [inline doc/architecture.md:22]: This section starts to get pretty wordy and hard to follow. Maybe we could break it up a bit. For example, show an example snippet of a Dockerfile. Just in general I think this section is chaining together too many MP/Aria Ops-specific concepts that are new to the reader, making it difficult to follow.

File: doc/architecture.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+endpoint served by the REST server to an executable command that will run as a separate 
+process inside the container. Importantly, the server appends two extra arguments to 
+each process call: An input and output named pipe. These are used for interprocess 
+communication.
+
+Because commands are run as a subprocess, the server is able to detect and handle many
+common errors (such as the process crashing without writing a result) without affecting
+the operation of the server.
+
+Importantly, this model enforces that adapters remain stateless. Because each call
```
Comment (MarkArbogast) [inline doc/architecture.md:43]: To help reduce complexity, I think this document should focus on the "what" and not go too much into the "why". I think the "why" is important here, but maybe we can find a way to sequester or move it to another page.

File: doc/architecture.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  adapter in an inconsistent state is reduced.
+ 
+Note that it is possible to pass data between collections by writing and reading to a 
+file on the container filesystem, but not encouraged. Any restarts of the adapter 
+instance (either by manual action or automatic migration) will create a new container, 
+and any data on the container filesystem will be lost.
+
+![Cloud Proxy Components running two Adapter Container Images](cloud-proxy-components.png)
+
+#### Commands.cfg
```
Comment (MarkArbogast) [inline doc/architecture.md:59]: I'm confused about this as a subheader. Although the document mentions it, we haven't previously indicated that `commands.cfg` is an important component, on the same level as the Adapter Container image. This section may be better served as its own page.

File: doc/architecture.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
+Architecture
```
Comment (MarkArbogast) [inline doc/architecture.md:1]: Overall, there's a lot of good information on this page. I think we should organize it more strategically and provide signposts at the beginning that tie everything together in a cohesive way.

File: doc/architecture.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+The user requires push access when building a Pack File, and the VMware Aria Operations
+Cloud Proxy needs anonymous pull access to the registry.
+
+Within a Pack File, the adapter container image is referenced by the registry location
+and image digest SHA.
+
+The following diagram shows how the processes for building and installing a Management 
+Pack work together with the registry.
+
+![Building and installing a Pack File](registry.png)
```
Comment (MarkArbogast) [inline doc/architecture.md:135]: There are too many things happening in this one image. Can we split it into two images, one for build, one for install?

**Review by MarkArbogast — CHANGES_REQUESTED** [review]

File: doc/architecture.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+results in a new subprocess being created, there is no way to save any state between 
+calls in the runtime environment. 
+Some benefits of a stateless adapter are:
+* Potential for (slow) memory leaks is reduced.
+* Potential for an adapter to return stale (cached) data is reduced.
+* Potential for an adapter to crash or hang due to a previous collection leaving the
+  adapter in an inconsistent state is reduced.
+ 
+Note that it is possible to pass data between collections by writing and reading to a 
+file on the container filesystem, but not encouraged. Any restarts of the adapter
```
[the remark this answers is not in the document]
Comment (kjrokos) [reply doc/architecture.md:53]: It is impossible in the runtime environment (i.e., preserving a cache in a python global variable). However, I think that this document is an appropriate place to discuss what workarounds are possible and the caveats associated with them, if a developer absolutely needs to save state somewhere.

File: doc/architecture.md (hunk 5 above)
Comment (kjrokos) [reply doc/architecture.md:59]: I think the info is appropriate here, but I've renamed and reorganized the section so it's more clear how it fits in

File: doc/architecture.md (hunk 7 above)
Comment (kjrokos) [reply doc/architecture.md:135]: What I really want to show here is the through line of the pack file and container image between the two processes. Maybe there's some details that aren't relevant, but it's a fine line between too much detail and oversimplification. For now, I've just added some additional text around it and renamed the section it's in to hopefully be clearer about what the diagram is trying to accomplish

File: doc/architecture.md (hunk 4 above)
Comment (kjrokos) [reply doc/architecture.md:43]: Yeah, I agree. I've removed it for now, but will try to find a better place for it elsewhere

File: doc/architecture.md (hunk 3 above)
Comment (kjrokos) [reply doc/architecture.md:22]: I added an example here, but otherwise I didn't change this section too much. I agree this is somewhat dense, but I don't know if I can avoid that in general for what I want this document to accomplish.

File: doc/architecture.md (hunk 1 above)
Comment (kjrokos) [reply doc/architecture.md:10]: We don't really have pages for the describe.xml or manifest.txt files to link to, but I've added links to the relevant content sections.

File: doc/architecture.md [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+ 
+### Adapter/Collection Code
+The Collection code is the main user-written code. It provides three functions:
+* 'Test': Determines if connection information entered by a user is valid.
+* 'Collect': Collect data from an external source and format it into an object structure 
+   which VMware Aria Operations can understand.
+* 'Get Endpoints': If the connection to the external source uses HTTPS, this method can
+   be used to tell VMware Aria Operations where to find certificates, and if they are
+   not signed by a Certificate Authority, present them to the user.
+Optionally, the adapter can provide a fourth method:
```
Comment (MarkArbogast) [inline doc/architecture.md:39]: Missing line break before "Optionally"
